felix/bpf-gpl/parsing.h

```diff
@@ -10,6 +10,44 @@
 #define PARSING_ALLOW_WITHOUT_ENFORCING_POLICY 2
 #define PARSING_ERROR -1
 
+#define MAX_EXTENSIONS 5
+
+static CALI_BPF_INLINE int parse_ipv6_extensions(struct cali_tc_ctx *ctx) {
+	struct ipv6_opt_hdr  ipv6_ext;
+	long header_length = ctx->ipheader_len;
+	__u8 next_header = ipv6_hdr(ctx)->nexthdr;
+	int i = 0;
+
+	for (i = 0; i < MAX_EXTENSIONS; i++) {
+		switch (next_header) {
+		case IPPROTO_HOPOPTS:
+		// Must be the first option, but at the moment, we only want to find
+		// transport layer, so we don't care about ordering.
+		case IPPROTO_ROUTING:
+		case IPPROTO_DSTOPTS:
+		case IPPROTO_AH:
+			if (bpf_skb_load_bytes(ctx->skb, skb_iphdr_offset() + header_length,
+						&ipv6_ext, sizeof(ipv6_ext))) {
+				CALI_DEBUG("Failed to load ipv6 extension header\n");
+				return PARSING_ERROR;
+			}
+
+			if (next_header == IPPROTO_AH)
+				header_length += (ipv6_ext.hdrlen * 4 + 8);
+			else
+				header_length += (ipv6_ext.hdrlen * 8 + 8);
+
+			next_header = ipv6_ext.nexthdr;
+			break;
+		default:
```

Comment:
How is the end of the header signalled? There is some sort of sentinel, right? This default would cut off at the first header we do not know. Right?

Reply:
The switch statement just checks for the IPv6 extensions, if the value is something other than the one we check, the code hits the default path, and exits. This function only calculated the length of IPv6 extension headers, and the rest will be handled in `tc_state_fill_from_nexthdr`. If the next header is one of `tcp`, `udp` or `icmp`, then we process it, otherwise we drop the protocols we don't recognise. Does it make sense?